File: commands/cash/destroy_cash.py
from argparse import Namespace, _SubParsersAction

from world import load_world, save_world


def run(args: Namespace) -> None:
    world = load_world()

    central_bank_id = args.central_bank_id
    amount = args.amount

    if amount <= 0:
        print("Amount must be greater than 0.")
        return

    if central_bank_id not in world["central_banks"]:
        print(f"Central bank does not exist: {central_bank_id}")
        return

    central_bank = world["central_banks"][central_bank_id]
    currency = central_bank["currency"]

    if central_bank["cash_vault"] < amount:
        print(
            f"Not enough physical cash in {central_bank_id}'s vault. "
            f"Available: {central_bank['cash_vault']} {currency}"
        )
        return

    central_bank["cash_vault"] -= amount
    central_bank["cash_issued"] -= amount

    save_world(world)

    print(f"Destroyed {amount} {currency} cash at {central_bank_id}.")
    print(f"{central_bank_id} cash_issued: {central_bank['cash_issued']} {currency}")
    print(f"{central_bank_id} cash_vault: {central_bank['cash_vault']} {currency}")


def register(subparsers: _SubParsersAction) -> None:
    parser = subparsers.add_parser(
        "destroy-cash", help="Destroy physical cash held in a central bank vault."
    )

    parser.add_argument("central_bank_id", help="Central bank ID, for example: ecb")

    parser.add_argument("amount", type=int, help="Amount of physical cash to destroy.")

    parser.set_defaults(func=run)

File: go/internal/commands/cash/destroy_cash.go
package cash

import (
	"fmt"
	"strconv"

	"mini-world-go/internal/commandlog"
	"mini-world-go/internal/commandrun"
	"mini-world-go/internal/domain"
	"mini-world-go/internal/world"

	"github.com/spf13/cobra"
)

func newDestroyCashCommand() *cobra.Command {
	return &cobra.Command{
		Use:   "destroy-cash central_bank_id amount",
		Short: "Destroy physical cash held in a central bank vault.",
		Args:  cobra.ExactArgs(2),
		RunE:  runDestroyCash,
	}
}

func runDestroyCash(cmd *cobra.Command, args []string) error {
	centralBankID := args[0]
	amount, err := strconv.Atoi(args[1])
	if err != nil {
		return fmt.Errorf("amount must be an integer")
	}

	w, err := world.Load()
	if err != nil {
		return err
	}

	if err := domain.DestroyCash(w, centralBankID, amount); err != nil {
		return commandrun.PrintBusinessError(err)
	}

	if err := commandrun.SaveWithHistory(w, "destroy-cash", args); err != nil {
		return err
	}

	centralBank := w.CentralBanks[centralBankID]
	currency := centralBank.Currency
	commandlog.Action("Destroyed %d %s cash at %s", amount, currency, centralBankID)
	commandlog.State("%s cash_issued: %d %s", centralBankID, centralBank.CashIssued, currency)
	commandlog.State("%s cash_vault: %d %s", centralBankID, centralBank.CashVault, currency)
	return nil
}
